src/metadataGeneration/typeResolver.ts

```diff
@@ -70,7 +70,7 @@ export class TypeResolver {
         return {
           dataType: 'union',
           types,
-        } as Tsoa.UnionType;
+        } as Tsoa.UnionType; // XXX why is this necessary?
```

Comment:
Please do not remove `as Tsoa.UnionType` or any similar stuff
This makes sure all props are set on the returned object.
If I want to return a specific type, I need to make sure all the additional stuff that type requires are added. The cast makes sure this is the case.

Reply:
Personally I would prefer typeguards over aliases. Like if we run this through a function that returns the type `input is Tsoa.Union` and then throw if that result is false, then we get a runtime check and more accurate types.

Reply:
> This makes sure all props are set on the returned object.

That's (unfortunately) not what this cast does. It's actually a [type *assertion*](https://www.typescriptlang.org/docs/handbook/basic-types.html#type-assertions), meaning that you're basically telling the compiler "Hey, I don't care what you think this is, from now it has this type".
If the types don't match the compiler will *sometimes* give a warning, but definitely not always (in this instance removing the `dataType` property didn't trigger any response from the compiler).

In this particular instance the issue was that the return type of the method was set to `Tsoa.Type`. `Tsoa.UnionType` is a subclass, but the compiler doesn't consider this (and rightfully so), throwing an error about properties that don't exist in `Tsoa.Type`. There are two solutions to this:
1. Add a type assertion (current situation)
2. Change the return type to `Tsoa.Type | Tsoa.UnionType`

I think option 2 is better because it more accurately represents the typing and most importantly **it actually checks the type**.
If the object you're returning doesn't match `Tsoa.Type` or `Tsoa.UnionType` it complains. If you're using a type assertion (telling it "this is definitely a `UnionType`) you won't get any errors.

Reply:
A third solution would be something like this:
```ts
const someType: Tsoa.SomeType = {
    // valid props for this type
};
return someType;
```
This avoids the need for a long union as return type and can be seen as more "readable" due to the explicit mention of what kind of type we're returning. It's however a little clunky in my opinion.

It'd be awesome if Typescript had a weaker kind of type assertion to ask the compiler to verify a type ("Can you check if this matches type X?" instead of "This is type X now"), but I'm not aware of such a feature.

Reply:
> That's (unfortunately) not what this cast does. It's actually a type assertion, meaning that you're basically telling the compiler "Hey, I don't care what you think this is, from now it has this type". If the types don't match the compiler will sometimes give a warning, but definitely not always (in this instance removing the dataType property didn't trigger any response from the compiler). If the types don't match the compiler will sometimes give a warning, but definitely not always (in this instance removing the dataType property didn't trigger any response from the compiler).

I'd expected it to force me to convert it to unknown first, if it does not, that's unfortunate.

> Change the return type to Tsoa.Type | Tsoa.UnionType

```ts
 Tsoa.Type | Tsoa.UnionType | Tsoa.ArrayType | Tsoa.EnumerateType | Tsoa.IntersectionType | Tsoa.ObjectLiteralType | Tsoa.ReferenceType
```
And this would change every time someone adds a type. I don't like that very much.

However, Dan has a lot more TS knowledge here so I'd gladly defer to him there.
Maybe it's a good idea to break up the entire method into several smaller ones.

Reply:
> I'd expected it to force me to convert it to unknown first, if it does not, that's unfortunate.

It surprised me at first too, it's really annoying. In *most* cases it asks you to convert first, but strangely enough not here.

> And this would change every time someone adds a type. I don't like that very much.

I'm not that big of a fan either. However, it allows for stricter type checking which I honestly prefer to having a single long union somewhere that needs to be updated every time a new 'type-type' is added (which shouldn't be that often anyway). We already have multiple of such unions, and I think that as long as it's properly documented what needs to be updated for new types it shouldn't cause too much hassle. Even without documentation I was able to figure it out pretty quickly when I was implementing tuples.
An even better solution would probably be to refactor to a type system that isn't based on these sorts of 'string literal unions', but that's a discussion for another day.

> Maybe it's a good idea to break up the entire method into several smaller ones.

I don't remember exactly what method we're talking about here, but there are definitely a couple of massive methods that could use this.

Reply:
Ideally, and if I were to rewrite this code today, I would have `Tsoa.Type` be a tagged union. There'd be no need for type assertions since you could only return type checked members of the union type.

Type assertions are sometimes necessary but generally to be avoided IMO. It would be quite easy to break this code by extending UnionType in some way since the type assertion will look over the fact that you're missing new or changed properties.

Even though it's uglier, I think a pattern of:

```
const myType: SomeType;
return myType;
```

will end up being safer.

Reply:
> Ideally, and if I were to rewrite this code today, I would have Tsoa.Type be a tagged union. There'd be no need for type assertions since you could only return type checked members of the union type.

Would this be a good change for a possible v3?

> Even though it's uglier, I think a pattern of: [...] will end up being safer.

This would be a good point to discuss in the style guide (even though there are objective safety benefits to this approach). I agree that type assertions should definitely be avoided as much as possible because they undermine the typesafety benefits of using Typescript. In my own projects I always add a comment to each type assertion to explain *why* this assertion is safe, to make sure others or myself don't get confused or undermine typesafety further.

Reply:
> Would this be a good change for a possible v3?

Probably. Knowing what I know now, type resolver is just kind of...wrong, but good enough. Redesigning the types around metadata is part of how we can fix it.

I think this particular thing is not so much style, but more of a system design thing. `Tsoa.Type` is an overly reductive return type because the callers of typeResolver actually care about the subtypes. Having to do the temp variable to avoid the assertion is more of a smell than anything.

Reply:
I 100% agree with you. Most of my changes are aimed at making the typeResolver feel less "wrong" (and still good enough), it probably won't be perfect but I hope it'll at least help.